### ISSUE TITLE
API via AJAX

### ISSUE THREAD

**Issue by distantnative** [getkart/kart#16]
A thought that just came across my mind but I can't grasp any real idea on this:

Within a theme (if not even the starterkit) it would be nice to be able to change e.g. the cart without reloading the site, e.g. deleting an item from the cart or increasing its quantity on the checkout site. This could be realized through some form of js-ajax functions. But then the core would need a API (JSON or something) for that as well, right? Or is this something the developers would built themselves then?

As I said, my brain is a lil fuzzy right now, so maybe it's dead simple and I just can't see it right now :D

**Comment by benseitz** [comment]
I agree adding items to the cart or changing the quantity should be done with AJAX instesd of a page reload.
I'm not really a JavaScript god, but with AngularJS you can ajax (sorry my brain is fuzzy too, I know ajax isn't a verb but you get what I mean) load the current page with post parameters and retrieve JSON.
The page would then check if the request is ajax (http://getkirby.com/docs/cheatsheet/request/ajax) and if so echo the data as json - else it would load the page itself.

So there's no need for a special AJAX API - you could build all that in the theme. But correct me if I'm wrong!

**Comment by plaidscreen** [comment]
I mentioned quickly while we were discussing the folder structure that I wanted a javascript core library, it's basically for this reason. I want to be able to provide the theme developers with Javascript functions to do stuff like `Kart.addItemToCart(...)` (I haven't thought about function definitions yet, that's another discussion.), and the `addItemToCart` function would interact with an API through AJAX, and the API would return JSON objects.

So, the core plugin would provide an API. That API could be extended by extension developers. And I would encourage the extension developers also to attach javascript helper functions, so even though `Kart.calculateShipping(...)` might not be part of the core, (because Shipping is an extension) the Shipping extension could implement that function (and the appropriate API routing) itself.

**Comment by benseitz** [comment]
That sounds great :) 
So plain JavaScript or jQuery (which would be really similar to Kirby API)?

**Comment by plaidscreen** [comment]
@benseitz Definitely jQuery. I try not to reinvent the wheel if I don't have to. :smile:

**Comment by distantnative** [comment]
Great, that's sounds just like I'd love to use it :smile:

**Comment by benseitz** [comment]
:+1:
